### PR TITLE
Added option to rename files in tools.copyTemplateFiles

### LINKED ISSUE #9: Add option to rename file when using tools.copyTemplateFiles
Now `tools.copyTemplateFiles` can be used as follows:
```js
tools.copyTemplateFiles( 
    [ 'README.md', 'CHANGES.md' ], 
    '/new/path',
    { '{{AppName}}': 'ckeditor5' } 
);
```
Sometimes there is a need to rename file when copying. For example, `.gitignore` files are renamed by npm when package is installed in `node_modules`. This way template of `.gitignore` must be stored under different name.
I propose to add option to pass object instead of string in sources array:
```js
tools.copyTemplateFiles( 
    [ { filePath: 'README_template.md', renameTo: 'README.md' }, 'CHANGES.md' ], 
    '/new/path',
    { '{{AppName}}': 'ckeditor5' } 
);
```

EDIT: changed fields of proposed object from `source` and `destination` to `filePath` and `renameTo`. Destination path is provided already as second param.

### PR DESCRIPTION
**Pull request by szymonkups**
Fixes #9. 
Added option to call `tools.copyTemplateFiles` with array containing object `{ filePath, renameTo }`:
```js
tools.copyTemplateFiles( [ { filePath: 'path/to/README.md.template', renameTo: 'README.md' ], '/new/path' );
```

### CODE REVIEW

File: lib/tools.js [hunk 1]
```diff
@@ -208,33 +208,46 @@ module.exports = {
...
 	 *		// Each occurrence of `{{appName}}` inside README.md and CHANGES.md will be changed to `ckeditor5`.
-	 * 		tools.copyTemplateFiles( [ 'README.md', 'CHANGES.md' ], '/new/path', { '{{AppName}}': 'ckeditor5' } );
+	 *		tools.copyTemplateFiles( [ '/path/to/README.md', '/path/to/CHANGES.md' ], '/new/path', { '{{AppName}}': 'ckeditor5' } );
 	 *
-	 * @param {Array} sources Source files.
+	 * Allows to change name of copied files:
+	 *
+	 *		tools.copyTemplateFiles( [ { filePath: 'path/to/README.md.template', renameTo: 'README.md' ], '/new/path' );
+	 *
+	 * @param {Array.<String|Object} sources Source files. Renaming source file after copying can be done by providing
```
Comment (maxbarnas) [inline lib/tools.js]: Is this correct? `Array.<String|Object`

Comment (szymonkups) [inline lib/tools.js]: It should be closed: `{Array.<String|Object>}

File: lib/tools.js [hunk 2]
```diff
@@ -208,33 +208,46 @@ module.exports = {
...
 		}
-		const regexp = new RegExp( regexps.join( '|' ), 'g' );
-		const replaceFunction = ( matched ) => replace[ matched ];
 
 		fs.ensureDirSync( destination );
 
-		sources.forEach( source => {
-			source = path.resolve( source );
+		sources.forEach( item => {
+			const rename = typeof item == 'object' && item.filePath && item.filePath;
```
Comment (maxbarnas) [inline lib/tools.js]: The `item.filePath` is twice here.

File: lib/tools.js [hunk 3]
```diff
@@ -208,33 +208,46 @@ module.exports = {
 	 * Copies source files into destination directory and replaces contents of the file using provided `replace` object.
 	 *
 	 *		// Each occurrence of `{{appName}}` inside README.md and CHANGES.md will be changed to `ckeditor5`.
-	 * 		tools.copyTemplateFiles( [ 'README.md', 'CHANGES.md' ], '/new/path', { '{{AppName}}': 'ckeditor5' } );
+	 *		tools.copyTemplateFiles( [ '/path/to/README.md', '/path/to/CHANGES.md' ], '/new/path', { '{{AppName}}': 'ckeditor5' } );
 	 *
-	 * @param {Array} sources Source files.
+	 * Allows to change name of copied files:
+	 *
+	 *		tools.copyTemplateFiles( [ { filePath: 'path/to/README.md.template', renameTo: 'README.md' ], '/new/path' );
```
Comment (Reinmar) [inline lib/tools.js]: The paths in the previous code sample are absolute. Shouldn't this be too?

File: tests/tools.js [hunk 4]
```diff
@@ -407,6 +407,31 @@ describe( 'utils', () => {
 				savePath = path.join( destination, path.basename( sources[ 1 ] ) );
 				sinon.assert.calledWithExactly( writeFileStub.secondCall, savePath, 'foo, bar, barfoo', 'utf8' );
 			} );
```
Comment (Reinmar) [inline tests/tools.js]: Missing new line.

File: lib/tools.js [hunk 5]
```diff
@@ -208,33 +208,46 @@ module.exports = {
...
+			const rename = typeof item == 'object' && item.filePath && item.renameTo;
+			const source = path.resolve( rename ? item.filePath : item );
+			const fileName = rename ? item.renameTo : path.basename( item );
 			let fileData = fs.readFileSync( source, 'utf8' );
-			fileData = fileData.replace( regexp, replaceFunction );
-			fs.writeFileSync( path.join( destination, path.basename( source ) ), fileData, 'utf8' );
+
+			if ( regexp ) {
+				fileData = fileData.replace( regexp, replaceFunction );
+			}
```
Comment (Reinmar) [inline lib/tools.js]: Do that pass through JSCS? I thought there must be an empty line after a block.

File: lib/tools.js (hunk 3 above)
Comment (Reinmar) [inline lib/tools.js]: And `}` is missing.

Comment (szymonkups) [inline lib/tools.js]: This should be absolute too.

File: lib/tools.js (hunk 5 above)
Comment (szymonkups) [inline lib/tools.js]: I will add a newline here, but JSCS is not giving any warring about this.

**Comment by Reinmar** [comment]
I've got a general question – why not a copy function where you specify the file that you want to copy and its destination (along with its name if you want).

We could use `copySync` from https://www.npmjs.com/package/fs-extra.

**Comment by szymonkups** [comment]
This method also replace some template strings on the fly, so probably this was the reason to read, modify and save the file instead of using `copy` method.

**Comment by Reinmar** [comment]
But we can still have it like this – `copyTemplateFile( inputPath, outputPath, replacements )`.

This method now became quite complicated. Both the params and its implementation because of that. There's really a lot of conditions inside it. Do you think it'll be possible to simplify it so it does one (actually two :D) thing in a simple way?

**Comment by szymonkups** [comment]
We can go with that. Its current implementation was strictly connected with `create-package` task. Now when the method is inside `ckeditor5-dev-utils`, we can make it more general and modify a little `create-package` task.

**Pull request merged**